<!DOCTYPE html>
<html lang='en'>
<head>
<title>fix-glob-dtype.patch\make\main - aports - Main aports tree
</title>
<meta name='generator' content='cgit v1.1'/>
<meta name='robots' content='index, nofollow'/>
<link rel='stylesheet' type='text/css' href='/cgit.css'/>
<link rel='shortcut icon' href='//alpinelinux.org/alpine-logo.ico'/>
<link rel='alternate' title='Atom feed' href='https://git.alpinelinux.org/cgit/aports/atom/main/make/fix-glob-dtype.patch?h=master' type='application/atom+xml'/>
<link rel='vcs-git' href='git://git.alpinelinux.org/aports' title='aports Git repository'/>
<link rel='vcs-git' href='https://git.alpinelinux.org/cgit/aports' title='aports Git repository'/>
</head>
<body>
<div id='cgit'><table id='header'>
<tr>
<td class='logo' rowspan='2'><a href='/cgit/'><img src='//wiki.alpinelinux.org/w/resources/assets/alogo.png' alt='cgit logo'/></a></td>
<td class='main'><a href='/cgit/'>index</a> : <a title='aports' href='/cgit/aports/'>aports</a></td><td class='form'><form method='get'>
<select name='h' onchange='this.form.submit();'>
<option value='1.10-stable'>1.10-stable</option>
<option value='1.9'>1.9</option>
<option value='2.0-stable'>2.0-stable</option>
<option value='2.1-stable'>2.1-stable</option>
<option value='2.2-stable'>2.2-stable</option>
<option value='2.3-stable'>2.3-stable</option>
<option value='2.4-stable'>2.4-stable</option>
<option value='2.5-stable'>2.5-stable</option>
<option value='2.6-stable'>2.6-stable</option>
<option value='2.7-stable'>2.7-stable</option>
<option value='3.0-stable'>3.0-stable</option>
<option value='3.1-stable'>3.1-stable</option>
<option value='3.2-stable'>3.2-stable</option>
<option value='3.3-stable'>3.3-stable</option>
<option value='3.4-stable'>3.4-stable</option>
<option value='3.5-stable'>3.5-stable</option>
<option value='3.6-stable'>3.6-stable</option>
<option value='3.7-stable'>3.7-stable</option>
<option value='master' selected='selected'>master</option>
</select> <input type='submit' value='switch'/></form></td></tr>
<tr><td class='sub'>Main aports tree
</td><td class='sub right'>gitolite</td></tr></table>
<table class='tabs'><tr><td>
<a href='/cgit/aports/about/'>about</a><a href='/cgit/aports/'>summary</a><a href='/cgit/aports/refs/'>refs</a><a href='/cgit/aports/log/main/make/fix-glob-dtype.patch'>log</a><a class='active' href='/cgit/aports/tree/main/make/fix-glob-dtype.patch'>tree</a><a href='/cgit/aports/commit/main/make/fix-glob-dtype.patch'>commit</a><a href='/cgit/aports/diff/main/make/fix-glob-dtype.patch'>diff</a><a href='/cgit/aports/stats/main/make/fix-glob-dtype.patch'>stats</a></td><td class='form'><form class='right' method='get' action='/cgit/aports/log/main/make/fix-glob-dtype.patch'>
<select name='qt'>
<option value='grep'>log msg</option>
<option value='author'>author</option>
<option value='committer'>committer</option>
<option value='range'>range</option>
</select>
<input class='txt' type='text' size='10' name='q' value=''/>
<input type='submit' value='search'/>
</form>
</td></tr></table>
<div class='path'>path: <a href='/cgit/aports/tree/'>root</a>/<a href='/cgit/aports/tree/main'>main</a>/<a href='/cgit/aports/tree/main/make'>make</a>/<a href='/cgit/aports/tree/main/make/fix-glob-dtype.patch'>fix-glob-dtype.patch</a></div><div class='content'>blob: 56c4e6c0505766df810578c0b7b0ff93919a1c19 (<a href='/cgit/aports/plain/main/make/fix-glob-dtype.patch'>plain</a>)
<table summary='blob content' class='blob'>
<tr><td class='linenumbers'><pre><a id='n1' href='#n1'>1</a>
<a id='n2' href='#n2'>2</a>
<a id='n3' href='#n3'>3</a>
<a id='n4' href='#n4'>4</a>
<a id='n5' href='#n5'>5</a>
<a id='n6' href='#n6'>6</a>
<a id='n7' href='#n7'>7</a>
<a id='n8' href='#n8'>8</a>
<a id='n9' href='#n9'>9</a>
<a id='n10' href='#n10'>10</a>
<a id='n11' href='#n11'>11</a>
<a id='n12' href='#n12'>12</a>
<a id='n13' href='#n13'>13</a>
<a id='n14' href='#n14'>14</a>
<a id='n15' href='#n15'>15</a>
<a id='n16' href='#n16'>16</a>
<a id='n17' href='#n17'>17</a>
<a id='n18' href='#n18'>18</a>
<a id='n19' href='#n19'>19</a>
<a id='n20' href='#n20'>20</a>
<a id='n21' href='#n21'>21</a>
<a id='n22' href='#n22'>22</a>
<a id='n23' href='#n23'>23</a>
<a id='n24' href='#n24'>24</a>
<a id='n25' href='#n25'>25</a>
<a id='n26' href='#n26'>26</a>
<a id='n27' href='#n27'>27</a>
<a id='n28' href='#n28'>28</a>
<a id='n29' href='#n29'>29</a>
<a id='n30' href='#n30'>30</a>
<a id='n31' href='#n31'>31</a>
</pre></td>
<td class='lines'><pre><code>musl 1.1.19 introduced the _DIRENT_HAVE_D_* macros in commit
1bc10ffeaa7c7ce44b3e214e02e302642511c7c7, triggering a bug in GNU make
that&apos;s present up to the current version:

http://git.savannah.gnu.org/cgit/make.git/tree/glob/glob.c?id=5d653b535a0e0e607c8be2016bcaadd6f326a698#n1333

The glob replacement they ship includes and uses internally a
GLOB_ONLYDIR flag, which is opportunistic and implemented via d_type
if _DIRENT_HAVE_D_TYPE is defined, and it fails to consider that
DT_LNK could be a symlink to a directory. This is an ancient version
of glob.c; the modern GNU glob in glibc rightly considers DT_LNK. Thus
the bug is not seen on glibc systems (where the glob replacement is
not used), but will be seen on any system where _DIRENT_HAVE_D_TYPE is
defined and where configure decides to replace the libc glob with
their GNU one.

http://www.openwall.com/lists/musl/2018/03/13/1

diff --git a/glob/glob.c b/glob/glob.c
index f3911bc..d6bb8ba 100644
<span style="color:#0057ae">--- a/glob/glob.c</span>
<span style="color:#000000; font-weight:bold">+++ b/glob/glob.c</span>
&#64;&#64; -1330,7 +1330,7 &#64;&#64; glob_in_dir (pattern, directory, flags, errfunc, pglob)
 		  /* If we shall match only directories use the information
 		     provided by the dirent call if possible.  */
 		  if ((flags &amp; GLOB_ONLYDIR)
<span style="color:#0057ae">-		      &amp;&amp; d-&gt;d_type != DT_UNKNOWN &amp;&amp; d-&gt;d_type != DT_DIR)</span>
<span style="color:#000000; font-weight:bold">+		      &amp;&amp; d-&gt;d_type != DT_UNKNOWN &amp;&amp; d-&gt;d_type != DT_DIR &amp;&amp; d-&gt;d_type != DT_LNK)</span>
 		    continue;
 #endif
 
</code></pre></td></tr></table>
</div> <!-- class=content -->
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-2804446-12");
pageTracker._trackPageview();
} catch(err) {}</script>
</div> <!-- id=cgit -->
</body>
</html>
